applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_strategy.h

```diff
@@ -461,61 +468,210 @@ class MPMStrategy
+
+            #pragma omp for
+            for(int i = 0; i < static_cast<int>(mpm_model_part.Elements().size()); ++i){
+
+                auto element_itr = mpm_model_part.Elements().begin() + i;
+
+                array_1d<double,3> xg = element_itr->GetValue(GAUSS_COORD);
+                typename BinBasedFastPointLocator<TDim>::ResultIteratorType result_begin = results.begin();
+
+                Element::Pointer pelem;
 
-        KRATOS_INFO_IF("MPM_Strategy", this->GetEchoLevel() > 1) << "Main Solve - End" <<std::endl;
+                // FindPointOnMesh find the element in which a given point falls and the relative shape functions
+                bool is_found = SearchStructure.FindPointOnMesh(xg, N, pelem, result_begin, MaxNumberOfResults, Tolerance);
+
+                if (is_found == true)
+                {
+                    pelem->Set(ACTIVE);
+                    element_itr->GetGeometry() = pelem->GetGeometry();
+                    if (m_GeometryElement == "Triangle")
+                    {
+                        element_itr->GetGeometry()[0].Set(ACTIVE);
+                        element_itr->GetGeometry()[1].Set(ACTIVE);
+                        element_itr->GetGeometry()[2].Set(ACTIVE);
+                        if (TDim ==3)
+                        {
+                            element_itr->GetGeometry()[3].Set(ACTIVE);
+                        }
+                    }
+                    else if(m_GeometryElement == "Quadrilateral")
+                    {
+                        element_itr->GetGeometry()[0].Set(ACTIVE);
+                        element_itr->GetGeometry()[1].Set(ACTIVE);
+                        element_itr->GetGeometry()[2].Set(ACTIVE);
+                        element_itr->GetGeometry()[3].Set(ACTIVE);
+                        if (TDim ==3)
+                        {
+                            element_itr->GetGeometry()[4].Set(ACTIVE);
+                            element_itr->GetGeometry()[5].Set(ACTIVE);
+                            element_itr->GetGeometry()[6].Set(ACTIVE);
+                            element_itr->GetGeometry()[7].Set(ACTIVE);
+                        }
+                    }
+                }
+                else{
+                    KRATOS_INFO("MPM_Strategy.SearchElement") << "WARNING: Search Element for Particle " << element_itr->Id()
+                        << " is failed. Geometry is cleared." << std::endl;
 
-		return 0.00;
+                    element_itr->GetGeometry().clear();
+                    element_itr->Reset(ACTIVE);
+                    element_itr->Set(TO_ERASE);
+                }
+            }
+        }
     }
 
+
+    /**
+     * @brief Function to perform expensive checks.
+     * @details It is designed to be called ONCE to verify that the input is correct.
+     */
+    int Check() override
+    {
+        KRATOS_TRY
+
+        for (ModelPart::ElementsContainerType::iterator it = mr_mpm_model_part.ElementsBegin();
```

Comment:
you could use a const auto loop

Reply:
@philbucher I am not sure about this one philipp. Can explain me a bit?

Reply:
the new way (c++11) would look like this:
https://en.cppreference.com/w/cpp/language/range-for
~~~
for (const auto& r_elem : mr_mpm_model_part.Elements()) {
...
}
~~~

I like this one more bcs it is shorter, I don't think there is a difference in efficieny.
Plus it is easy to treat the objects you are looping as `const`

Just a suggestion, since you also want to learn sth new :)
Btw this also works nicely for other stl-containers. Especially you don't have to declare the type when you use auto

Reply:
Thanks, @philbucher! Sorry for being a bit slow for the response. But apparently, since the function `Check` is not a `const` function, I can't use `const` there.

Reply:
I see
Then the following would work (not saying that you should change it, just for you to know)
~~~
for (auto& r_elem : mr_mpm_model_part.Elements()) {
...
}
~~~

Reply:
Did it already. Can you recheck the last commit?